lls/cache.c

```diff
@@ -0,0 +1,468 @@
+	}
+}
+
+static void
+lls_update_subscribers(struct lls_record *record)
+{
+	unsigned int i;
+	for (i = 0; i < record->num_holds; i++) {
+		int call_again = false;
+
+		record->holds[i].cb(&record->map, record->holds[i].arg,
+			LLS_REPLY_RESOLUTION, &call_again);
+
+		if (call_again)
+			continue;
+
+		/* Delete hold; keep all holds in beginning of array. */
+		record->num_holds--;
+		if (i < record->num_holds) {
+			memcpy(&record->holds[i],
+				&record->holds[record->num_holds],
+				sizeof(record->holds[i]));
+			/*
+			 * This cancels out the update of the for loop so we
+			 * can redo update of hold at position @i, if needed.
+			 */
+			i--;
+		}
+	}
+}
+
+static int
+lls_add_record(struct lls_cache *cache, const uint8_t *ip_be)
+{
+	int ret = rte_hash_add_key(cache->hash, ip_be);
+	if (unlikely(ret == -EINVAL || ret == -ENOSPC)) {
+		char ip_buf[cache->key_str_len];
+		char *ip_str = cache->ip_str(cache, ip_be,
+			ip_buf, cache->key_str_len);
+		RTE_LOG(ERR, HASH, "%s, could not add record for %s\n",
+			ret == -EINVAL ? "Invalid params" : "No space",
+			ip_str == NULL ? cache->name : ip_str);
+	} else
+		RTE_ASSERT(ret >= 0);
+	return ret;
+}
+
+static void
+lls_del_record(struct lls_cache *cache, const uint8_t *ip_be)
+{
+	int32_t ret = rte_hash_del_key(cache->hash, ip_be);
+	if (unlikely(ret == -ENOENT || ret == -EINVAL)) {
+		char ip_buf[cache->key_str_len];
+		char *ip_str = cache->ip_str(cache, ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, HASH, "%s, record for %s not deleted\n",
+			ret == -ENOENT ? "No map found" : "Invalid params",
+			ip_str == NULL ? cache->name : ip_str);
+	}
+}
+
+static void
+lls_process_hold(struct lls_config *lls_conf, struct lls_hold *hold)
+{
+	struct lls_cache *cache = hold->cache;
+	struct lls_record *record;
+	int ret = rte_hash_lookup(cache->hash, hold->ip_be);
+
+	if (ret == -ENOENT) {
+		ret = lls_add_record(cache, hold->ip_be);
+		if (ret < 0)
+			return;
+
+		record = &hold->cache->records[ret];
+		record->map.stale = true;
+		memcpy(record->map.ip_be, hold->ip_be, cache->key_len);
+		record->ts = time(NULL);
+		RTE_ASSERT(record->ts >= 0);
+		record->holds[0] = *hold;
+		record->num_holds = 1;
+
+		/* Try to resolve record using broadcast. */
+		lls_send_request(lls_conf, cache, hold->ip_be, NULL);
+
+		if (lls_conf->debug)
+			lls_cache_dump(cache);
+		return;
+	} else if (unlikely(ret == -EINVAL)) {
+		char ip_buf[cache->key_str_len];
+		char *ip_str;
+		ip_str = cache->ip_str(cache, hold->ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, HASH,
+			"Invalid params, could not get %s map; hold failed\n",
+			ip_str == NULL ? cache->name : ip_str);
+		return;
+	}
+
+	RTE_ASSERT(ret >= 0);
+	record = &hold->cache->records[ret];
+
+	if (!record->map.stale) {
+		int call_again = false;
+		/* Alert requester this map is ready. */
+		hold->cb((const struct lls_map *)&record->map, hold->arg,
+			LLS_REPLY_RESOLUTION, &call_again);
+		if (!call_again)
+			return;
```

Comment:
The intent is to only dump the cache when there's a scan or when there's a change. In this case, the block just gets the resolution but no change to the cache is made.